sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/WebExceptionRetryPolicy.java

```diff
@@ -41,79 +37,73 @@ public WebExceptionRetryPolicy(RetryContext retryContext) {
+                            "isAddressRefresh = {}, shouldForcedAddressRefresh = {}, " +
+                            "shouldForceCollectionRoutingMapRefresh = {}",
+                        this.locationEndpoint, this.request.getOperationType(), this.retryCount,
+                        this.request.isAddressRefresh(),
+                        this.request.shouldForceAddressRefresh(),
+                        this.request.forceCollectionRoutingMapRefresh);
 
-            return Mono.just(ShouldRetryResult.retryAfter(Duration.ofSeconds(retryDelay)));
-        }
 
-        if (!WebExceptionUtility.isWebExceptionRetriable(e)) {
-            // Have caller propagate original exception.
-            this.durationTimer.stop();
-            return Mono.just(ShouldRetryResult.noRetryOnNonRelatedException());
+                return Mono.just(ShouldRetryResult.retryAfter(Duration.ofSeconds(delayInSeconds)));
+            }
         }
 
-        logger.warn("Received retriable web exception, will retry", e);
-        return Mono.just(ShouldRetryResult.retryAfter(Duration.ofSeconds(retryDelay)));
+
+        logger
+            .warn(
+                "WebExceptionRetryPolicy() No retrying on un-retryable exceptions on endpoint {}, operationType = {}, count = {}, " +
+                    "isAddressRefresh = {}",
+                this.locationEndpoint,
+                this.request.getOperationType(),
+                this.retryCount,
+                this.request.isAddressRefresh());
+
+
+        this.durationTimer.stop();
+        return Mono.just(ShouldRetryResult.noRetryOnNonRelatedException());
     }
 
     @Override
     public RetryContext getRetryContext() {
         return this.retryContext;
     }
 
+    // This method should only be called once
     public void onBeforeSendRequest(RxDocumentServiceRequest request) {
         this.request = request;
         this.isReadRequest = request.isReadOnlyRequest();
         this.timeoutPolicy = HttpTimeoutPolicy.getTimeoutPolicy(request);
-        this.isOutOfRetries = isOutOfRetries();
-
-        // Setting the current responseTimeout and retryDelay using the timeout policy being used
-        // and then increasing the counter for retries.
-        if (!isOutOfRetries) {
-            ResponseTimeoutAndDelays current = timeoutPolicy.getTimeoutAndDelaysList().get(this.retryCount);
-            this.request.setResponseTimeout(current.getResponseTimeout());
-            this.retryDelay = current.getDelayForNextRequestInSeconds();
-        }
-        this.retryCount++;
+
+        // set the initial response timeout
+        this.request.setResponseTimeout(timeoutPolicy.getTimeoutAndDelaysList().get(0).getResponseTimeout());
         this.locationEndpoint = request.requestContext.locationEndpointToRoute;
     }
 
     private boolean isOutOfRetries() {
         return this.retryCount >= this.timeoutPolicy.totalRetryCount();
```

Comment:
I think the isOutOfRetries should  `this.retryCount >= this.timeoutPolicy.totalRetryCount()-1;`

Reply:
Updated code